[AOTI] Store OpOverload in ir.ExternKernel

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120714
* #120513
* __->__ #120629
* #120592
* #119987

Summary: Currently the logics for filling the default value for optional arguments are scattered in several places. By storing OpOverload in the base ExternKernel class, we can simplify codegen_kwargs, and this is a preparation step for enabling the torchgen-ed C shim. The default value filling logic for FallbackKernel can also be simplified, but that can come later.

Differential Revision: [D54258089](https://our.internmc.facebook.com/intern/diff/D54258089)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames

desertfire committed:
[AOTI] Store OpOverload in ir.ExternKernel

Summary: Currently the logics for filling the default value for optional arguments are scattered in several places. By storing OpOverload in the base ExternKernel class, we can simplify the logics, and this is a preparation step for enabling the torchgen-ed C shim.

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Store OpOverload in ir.ExternKernel"


Summary: Currently the logics for filling the default value for optional arguments are scattered in several places. By storing OpOverload in the base ExternKernel class, we can simplify the logics, and this is a preparation step for enabling the torchgen-ed C shim.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54258089).

desertfire committed:
Update on "[AOTI] Store OpOverload in ir.ExternKernel"


Summary: Currently the logics for filling the default value for optional arguments are scattered in several places. By storing OpOverload in the base ExternKernel class, we can simplify the logics, and this is a preparation step for enabling the torchgen-ed C shim.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

Differential Revision: [D54258089](https://our.internmc.facebook.com/intern/diff/D54258089)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54258089).

desertfire committed:
Update on "[AOTI] Store OpOverload in ir.ExternKernel"


Summary: Currently the logics for filling the default value for optional arguments are scattered in several places. By storing OpOverload in the base ExternKernel class, we can simplify codegen_kwargs, and this is a preparation step for enabling the torchgen-ed C shim. The default value filling logic for FallbackKernel can also be simplified, but that can come later.

Differential Revision: [D54258089](https://our.internmc.facebook.com/intern/diff/D54258089)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54258089).

desertfire committed:
Update on "[AOTI] Store OpOverload in ir.ExternKernel"


Summary: Currently the logics for filling the default value for optional arguments are scattered in several places. By storing OpOverload in the base ExternKernel class, we can simplify codegen_kwargs, and this is a preparation step for enabling the torchgen-ed C shim. The default value filling logic for FallbackKernel can also be simplified, but that can come later.

Differential Revision: [D54258089](https://our.internmc.facebook.com/intern/diff/D54258089)

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Store OpOverload in ir.ExternKernel"


Summary: Currently the logics for filling the default value for optional arguments are scattered in several places. By storing OpOverload in the base ExternKernel class, we can simplify codegen_kwargs, and this is a preparation step for enabling the torchgen-ed C shim. The default value filling logic for FallbackKernel can also be simplified, but that can come later.

Differential Revision: [D54258089](https://our.internmc.facebook.com/intern/diff/D54258089)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54258089).

chenyang78 commented on torch/_inductor/ir.py:
```diff
@@ -3870,50 +3873,91 @@ def apply_constraint(self):
     def codegen_const_args(self):
         return map(V.graph.wrapper_code.val_to_arg_str, self.constant_args)
 
+    def collect_arg_kwarg_properties(self):
```
It's still not clear to me where we should invoke ```collect_arg_kwarg_properties```. I see that we call it from multiple places like ```codegen_args``` and ```__init__``` functions. So, I guess my questions is - why is it not sufficient to just call it from ```__init__```? If it's not sufficient due to some side-effects happening between the ```__init__``` routine and ```codegen_args```, could we only call ```collect_arg_kwarg_properties``` from ```codegen_args``` instead? 

Unifying the call sites could help understand the code logic. Thanks!

desertfire replied:
The main reason I chose to do it this way is because the base class, ExternKernel, is derived by many subclasses, so adding an extra field to the base class may require chances to all the subclass constructors. But you made a valid point. Let me update.

desertfire committed:
Update on "[AOTI] Store OpOverload in ir.ExternKernel"


Summary: Currently the logics for filling the default value for optional arguments are scattered in several places. By storing OpOverload in the base ExternKernel class, we can simplify codegen_kwargs, and this is a preparation step for enabling the torchgen-ed C shim. The default value filling logic for FallbackKernel can also be simplified, but that can come later.

Differential Revision: [D54258089](https://our.internmc.facebook.com/intern/diff/D54258089)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54258089).

chenyang78 reviewed (approved):
LGTM. Thanks!
